gregBorchers/HW6Problem2.js

```diff
@@ -0,0 +1,245 @@
+    if (!claim) console.error(message);
+}
+
+console.log("\n--- START of Simba testing-----");
+
+var simba = new Animal("Simba")
+console.log(simba);
+console.log(simba.move());
+console.log(simba.hasWings);
+console.log(simba instanceof Animal);
+console.log(simba instanceof Bird);
+console.log(simba instanceof Penguin );
+console.log(simba instanceof Fish);
+
+console.log("\n--- START of Nemo testing-----");
+var nemo = new Fish("Nemo");
+console.log(nemo);
+console.log(nemo.move());
+console.log(nemo.hasWings);
+console.log(nemo instanceof Animal);
+console.log(nemo instanceof Bird);
+console.log(nemo instanceof Penguin);
+console.log(nemo instanceof Fish);
+
+console.log("\n--- START of Lulu testing-----");
+var birdo = new Bird("Lulu");
+console.log(birdo);
+console.log(birdo.move());
+console.log(birdo.hasWings);
+console.log(birdo instanceof Animal);
+console.log(birdo instanceof Bird);
+console.log(birdo instanceof Penguin);
+console.log(birdo instanceof Fish);
+
+console.log("\n--- START of Pengo testing-----");
+var pengo = new Penguin("Pengo");
+console.log(pengo);
+console.log(pengo.move());
+console.log(pengo.hasWings);
+console.log(pengo instanceof Animal);
+console.log(pengo instanceof Bird);
+console.log(pengo instanceof Penguin);
+console.log(pengo instanceof Fish);
+
+assert( pengo.name === 'Pengo', "pengo.name FAIL");
+assert ( pengo.move() === 'swim', 'pengo.swim FAIL');
+assert ( pengo.hasWings === true, 'pengo.hasWings FAIL');
+assert ( pengo instanceof Penguin === true, 'FAIL pengo instanceof Penguin');
+assert ( pengo instanceof Animal === true, 'FAIL pengo instanceof Animal');
+assert ( pengo instanceof Bird ===  true, 'FAIL pengo instanceof Bird');
+
+/**
+b)** Create a class _Egg_, whose instances have one method, _hatch(name)_, 
+  which returns a new instance (named _name_) of the same species which laid the egg.
+Assume that every Animal can lay an egg with an instance method _layEgg()_ 
+  which creates a new Egg instance.
+Try to solve this without subclassing Egg and 
+  without implementing _layEgg_ and _hatch_ more than once.
+*/
+
+//_________________________________________________________________ EGG
+// 
+var Egg = (function() {
+	function Egg(speciesConstructor) {
+		this.speciesConstructor = speciesConstructor;
+
+	}
+	
+	Egg.prototype.hatch = function(babyName) {
+		// TOFIX this does not fully populate the prototype chain, although the Ctor seems correct.. ?
+		return new this.speciesConstructor(babyName);
+	};
+		
+	return Egg;
+
+})();
+
+
+//_________________________________________________________________ TESTING for 2b
+// 
+
+function assert(claim, message) {
+    if (!claim) console.error(message);
+}
+
+console.log("\n--- START of Pengo+Penglet Egg testing-----");
+
+var pengo = new Penguin("Pengo");
+var egg = pengo.layEgg();
+//egg.constructor === Egg; //true
+var baby = egg.hatch("Penglet");
+//baby instanceof Penguin; //true
+
+console.log("Penglet obj returns " + baby);
```

Comment:
Handy tip: for tests like this, if you break the `console.log` into two lines, so that the object is by itself, you'll be able to examine the object in your browser&mdash;at least, that's the case in Chrome.